### ISSUE TITLE
Krea2 LoRA trained locally on Mac with AI-toolkit 0.10.19 has all lora_B tensors at zero

### ISSUE THREAD

**Issue by jcharlesmarier-maker** [ostris/ai-toolkit#925]
Hi,

I’m having an issue when training a Krea2 LoRA locally on my Mac with AI-toolkit.

I trained two LoRAs using the same settings:

1. One LoRA was trained online and works correctly in ComfyUI.
2. The other LoRA was trained locally on my Mac and does not work properly in ComfyUI, even though I used the same training settings.

After comparing both `.safetensors` files, the structure looks almost identical:

* Same base model: `krea2`
* Same trigger word: `denisdrew`
* Same LoRA rank: 32
* Same number of tensors: 512
* Same number of LoRA modules: 256
* Same layer names / keys

The main difference is in the actual trained weights.

The working online LoRA was created with:

* AI-toolkit version: `0.10.18`
* File: `denis_drew_krea2_v01_000003000.safetensors`
* Training steps: 3000
* The `lora_A` and `lora_B` tensors both contain non-zero values.

The local Mac LoRA was created with:

* AI-toolkit version: `0.10.19`
* File: `denis_drew_krea2_v01_000002750.safetensors`
* Training steps: 2750
* All `lora_B` tensors appear to be entirely zero.

Because all `lora_B` tensors are zero, the LoRA seems to have almost no effect when loaded in ComfyUI. The file loads without an obvious error, but the trained character/style is not applied.

My setup:

* macOS
* Apple Silicon Mac
* AI-toolkit local version: `0.10.19`
* Base model: Krea2
* Training done locally through AI-toolkit

It looks like the local Mac training process may not be updating or saving the `lora_B` weights correctly. Could this be a regression in `0.10.19`, or possibly an Apple Silicon / MPS training issue?

Please let me know if there is anything specific I should test, such as trying version `0.10.18`, changing precision settings, disabling MPS, or checking a specific training option.

Thank you!

**Comment by robandiz25** [comment]
Hitting the exact same issue here, but on Windows/CUDA (RTX 5090, CUDA 13.0) — not Mac/MPS — so this doesn't look like an Apple Silicon-specific problem.

Trained a character LoRA on a Krea2-arch checkpoint (community diffusers-format finetune, converted to native key naming before training — conversion verified correct via clean strict state_dict load, no shape/key errors). Training completed normally: 2000+ steps, healthy decreasing loss (~1.9 → ~0.8 range), no NaN, no errors, checkpoints saved fine.

Config: `arch: krea2`, `type: lora`, rank/alpha 16/16, adamw8bit, lr 1e-4, EMA on, bf16, no quantization, layer_offloading on (needed for VRAM headroom on this checkpoint).

Loaded the resulting LoRA in ComfyUI — patches attach cleanly (log confirms "256 patches attached", matching the "create LoRA for U-Net: 256 modules" count from training). But the LoRA has zero visible effect at any strength (tested 1.0 and 2.0) — not even the usual overcooked/distorted look you'd expect from an over-strength LoRA, just literally the base model's output.

Dumped the actual tensor norms to check:

```
diffusion_model.blocks.0.attn.wq.lora_A.weight: shape=(16, 6144) norm=2.311796
diffusion_model.blocks.0.attn.wq.lora_B.weight: shape=(6144, 16) norm=0.000000
```

Every `lora_B` tensor I sampled across multiple layers (attn.wq/wk/wv/wo/gate) is exactly `0.000000` — `lora_A` is normally distributed (matches what you'd expect from random init), but `B` never moved from its zero-initialization despite 2000+ training steps and a normal-looking loss curve.

For comparison, a working Krea2 LoRA I trained previously (different checkpoint) on the same machine/setup shows `lora_B` norms in the 0.3–1.2 range at the same layer, same step count — so this looks like a real regression in how gradients reach `lora_B` for this arch, not a platform-specific fluke. Happy to share the full training config or run additional diagnostics if useful.

**Comment by jcharlesmarier-maker** [comment]
Thanks for sharing this — this is very useful because it confirms the issue is not necessarily Apple Silicon / MPS-specific.
In my case, the first broken Krea2 LoRA checkpoints had the same symptom: all lora_A.weight tensors were non-zero, but every lora_B.weight tensor was exactly zero. The LoRA would load in ComfyUI, but had no visible effect.
My original failing setup was on Mac/MPS with bf16, qfloat8 quantization, and the automagic optimizer. The training log showed loss is nan, lr: 0.0e+00, and the resulting LoRA had all lora_B tensors stuck at zero.
After switching to a safer config — float32, adamw, no quantization, no text encoder training, and training on krea/Krea-2-Raw — the issue disappeared on my side. The new checkpoints now have all lora_B.weight tensors non-zero, no NaN/Inf values, and the LoRA has a visible effect in ComfyUI.
So there may be two related failure modes:
A stability issue where training falls into NaN and produces zero-effect LoRA checkpoints.
A deeper Krea2 LoRA issue where lora_B never updates even though the loss appears normal, as reported here on CUDA.
It would be very helpful if the trainer could log or assert the gradient norms for *.lora_B.weight during training, or warn when a saved LoRA has all lora_B tensors still at zero. That would make this issue much easier to diagnose.
In my tests, checking the saved .safetensors directly was the only reliable way to confirm whether the LoRA actually learned anything.

**Comment by codered-42** [comment]
same issue for me, but for concept slider training, a summary of a claude-made workaround : 

#---------------------
Krea2 Concept Slider issue

error during Krea2 Concept Slider training on Windows with an RTX 4090 (24 GB). The first loss looks normal, but then the loss produces NaNs or zeros in the LoRA B layer. Because of that the saved LoRA either does nothing or has NaN weights and cannot work.

 Fix and suggestions

the problem is probably Krea2's bf16 attention together with gradient checkpointing and layer offloading. I also tested regular AdamW and disabled model quantization, so it does not look like an AdamW8bit or quantization-only problem.

The temporary fix uses a safer, chunked attention fallback during Concept Slider training. This works: the LoRA B weights are now nonzero and the LoRA actually trains. But it cannot use the normal fast attention backend, so it is really slow (slower , slowest...).

For a proper fix, it would probably need a fast and stable attention implementation, for example FlashAttention that supports Krea2's attention layout, or a fix for the current PyTorch/cuDNN attention backward path.

#---------------------

For anyone who wants to try this fix, here is a prompt for claude:
"""
In the Ostris AI Toolkit repo, fix Krea2 Concept Slider training where `lora_B` gradients become NaN or remain zero when using bf16 + gradient checkpointing + layer offloading.

Relevant files:
- `extensions_built_in/concept_slider/ConceptSliderTrainer.py`
- `extensions_built_in/diffusion_models/krea2/src/mmdit.py`

The issue is likely in Krea2 GQA attention backward: Krea2 uses 48 query heads and 12 key/value heads. The normal PyTorch/cuDNN SDPA backend gives a finite forward loss but produces invalid LoRA gradients during the two Concept Slider passes (`+1` and `-1` LoRA multipliers).

Implement a training-only stable attention path:
1. Add an opt-in flag on Krea2 attention modules, default `False`.
2. When enabled and gradients are active, use `torch.nn.functional.scaled_dot_product_attention` with the Math backend, not cuDNN/Flash/Efficient.
3. Process query tokens in chunks (default 128 tokens, configurable with `KREA2_TRAIN_ATTN_QUERY_CHUNK_SIZE`) so it fits in 24 GB VRAM.
4. Wrap every attention chunk in `torch.utils.checkpoint.checkpoint(..., use_reentrant=False)` to control memory.
5. Add `SingleStreamDiT.enable_stable_training_attention()` to enable this only on its attention layers.
6. Call that method only from Concept Slider training setup for `arch == "krea2"`. Do not slow down normal Krea2 LoRA training.
7. After backward, verify every Krea2 `lora_up.weight` / `lora_B` gradient is registered, finite, and nonzero. After the optimizer step, verify the B weights are finite and no longer all zero. Raise a clear error otherwise.
8. Keep the normal backend selection unchanged when stable mode is disabled.

Validate with `py_compile`, and add or run a small attention test showing finite output plus finite Q/K/V gradients for the 48Q/12KV GQA layout.
"""
